File: 2017/src/main/scala/day02/day2.scala
package day02

import scala.util.{Try, Success, Failure, Using}
import scala.io.Source

def parseInput(input: List[String]) = input.map(line => {
    raw"(\d+)".r.findAllIn(line).map(_.toInt).toVector.sorted
})

def evaluatorOne(input: List[Vector[Int]]): Int = {
    return input.map(it => it.last - it.head).sum
}

def evaluatorTwo(input: List[Vector[Int]]): Int = {
    return input.flatMap(values => {
        (for { 
            smaller <- values
            larger <- values.reverseIterator.takeWhile(_ >= 2*smaller)
            if larger % smaller == 0 
        } yield larger / smaller)
    }).sum
}

def readLinesFromFile(filePath: String): Try[List[String]] =
    Using(Source.fromResource(filePath))(_.getLines().toList)

def hello(): Unit = {
    readLinesFromFile("day02.txt") match {
        case Success(lines) => {
            val input = parseInput(lines)
            println(s"Part One: ${evaluatorOne(input)}")
            println(s"Part Two: ${evaluatorTwo(input)}")
        }
        case Failure(exception) => {
            println(s"Error reading file: ${exception.getMessage}")
        }
    }
}

File: intcode/src/main/kotlin/day02/day2.kt
package day02

import java.io.File
import intcode.IntCodeMachine

fun execIntCode(icm: IntCodeMachine, noun: Long, verb: Long): Long {
    icm.reset()
    icm.memory[1] = noun
    icm.memory[2] = verb
    icm.run()
    return icm.memory[0]
}

fun evaluatorOne(line: String) = execIntCode(IntCodeMachine(line), 12, 2)

fun evaluatorTwo(line: String): Int {
    val icm = IntCodeMachine(line)
    var sum = 0

    while(true) {
        for(verb in 0..sum) {
            val noun = sum - verb
            val res = execIntCode(icm, noun.toLong(), verb.toLong())
            if (res.toInt() == 19690720) return 100 * noun + verb
        }
        sum++
    }
}

fun readLinesFromFile(filePath: String): Result<List<String>> =
    runCatching { File(filePath).readLines() }

fun main() {
    readLinesFromFile("src/main/resources/day02.txt")
        .onSuccess {
            val line = it.first()
            println("Part One: ${evaluatorOne(line)}")
            println("Part Two: ${evaluatorTwo(line)}")
        }
        .onFailure {
            println("Error reading file: ${it.message}")
        }
}